index.html

```diff
@@ -2687,13 +2687,94 @@ <h1>
 This section defines the inputs and outputs of the <a>DID resolution</a> and
 <a>DID URL dereferencing</a> processes. The exact implementation of these
 functions is out of scope for this specification, but some considerations for
-implementors are available in [[?DID-RESOLUTION]].
+implementors are discussed in [[?DID-RESOLUTION]].
+    </p>
+
+    <p>
+All conformant <a>DID resolvers</a> MUST implement the <a>DID resolution</a> function
+for at least one <a>DID method</a> and MUST be able to return a <a>DID document</a> in 
+at least one conformant representation.
     </p>
 
     <section>
         <h2>
 DID Resolution
         </h2>
+        <p>
+The <a>DID resolution</a> function resolves a <a>DID</a> into a <a>DID document</a>
+by using the "Read" operation of the applicable <a>DID method</a>. (See <a href="#read-verify"></a>.)
+The details of how this process is accomplished are outside the scope of this
+specification, but all conformant implementations MUST implement a function in the following form:
+        </p>
+
+        <p><code>
+resolve ( did, did-resolution-input-metadata ) <br>
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-&gt; ( did-resolution-metadata, did-document-stream, did-document-metadata )
+        </code></p>
+
+        <p>
+The input variables of this function MUST be as follows:
+        </p>
+
+        <dl>
+            <dt>
+did
+            </dt>
+            <dd>
+A conformant <a>DID</a> as a single string. This is the <a>DID</a> to resolve.
```

Comment:
Do we want to specify whether it MUST be a DID URI or a DID URL here? In general, I've operated under the assumption resolution accepts only DID URIs and dereferencing accepts DID URLs.

Reply:
I don't think we're defining "DID URI" anywhere? Resolution takes a DID, dereferencing accepts a DID URL.

Reply:
Resolution always takes a DID.

Reply:
Just to clarify terminology first, when you're saying DID, do you mean just `did:example:123`? That's what I meant by DID URI, so I think we'd be in alignment then. However, I found a case that would conflict with my original assumption. The case where we're referring to a specific version of a did document with the version query parameter could use some clarity for this.

For example, if I used `did:example:123?version=1` would I need to parse `version=1` from the identifier and pass that via the options then or could I pass the full DID URL with the version query parameter to the resolution function?

Reply:
@kdenhartog 

> would I need to parse `version=1` from the identifier and pass that via the options then or could I pass the full DID URL with the version query parameter to the resolution function?

You can only pass a DID to `resolve()`. But you can pass a DID URL to `dereference()`. In that latter case, the DID parameters are parsed from the DID URL and then passed as `did-resolution-input-metadata` (aka "options") to `resolve()`, exactly like you said.

Example:

`dereference(did-url="did:example:123?version=1", did-dereference-input-metadata=null)`

will internally call

`resolve(did="did:example:123", did-resolution-input-metadata=[version="1"])`

See the first step in the [DID URL Dereferencing algorithm section of the DID Resolution spec](https://w3c-ccg.github.io/did-resolution/#dereferencing-algorithm), which says this:

_All generic DID parameters as well as method-specific DID parameters of the input DID URL MUST be passed as input metadata properties to the DID Resolution algorithm._